go/oasis-node/cmd/debug/txsource/workload/runtime.go

```diff
@@ -0,0 +1,257 @@
+		// expected.
+		if rsp.Success != nil {
+			return fmt.Errorf("expected nil response, got: '%s'", rsp.Success)
+		}
+	}
+
+	return nil
+}
+
+func (r *runtime) submitRuntimeRquest(ctx context.Context, rtc runtimeClient.RuntimeClient, req interface{}) (*runtimeTransaction.TxnOutput, error) {
+	var rsp runtimeTransaction.TxnOutput
+	rtx := &runtimeClient.SubmitTxRequest{
+		RuntimeID: r.runtimeID,
+		Data:      cbor.Marshal(req),
+	}
+
+	r.logger.Debug("submitting request",
+		"request", req,
+	)
+
+	// Wait for a maximum of 30 seconds as invalid submissions may block
+	// forever.
+	submitCtx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	out, err := rtc.SubmitTx(submitCtx, rtx)
+	if err != nil {
+		cancel()
+		return nil, fmt.Errorf("failed to submit runtime transaction: %w", err)
+	}
+	cancel()
+
+	if err = cbor.Unmarshal(out, &rsp); err != nil {
+		return nil, fmt.Errorf("malformed tx output from runtime: %w", err)
+	}
+	if rsp.Error != nil {
+		return nil, fmt.Errorf("runtime tx failed: %s", *rsp.Error)
+	}
+
+	return &rsp, nil
+}
+
+func (r *runtime) doInsertRequest(ctx context.Context, rng *rand.Rand, rtc runtimeClient.RuntimeClient, existing bool) error {
+	key := r.generateVal(rng, existing)
+	value := r.generateVal(rng, false)
+
+	// Submit request.
+	rsp, err := r.submitRuntimeRquest(ctx, rtc, &runtimeTransaction.TxnCall{
+		Method: "insert",
+		Args: struct {
+			Key   string `json:"key"`
+			Value string `json:"value"`
+		}{
+			Key:   key,
+			Value: value,
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("submit insert request failed: %w", err)
+	}
+	r.logger.Debug("insert request success",
+		"response", rsp,
+	)
+
+	if err := r.validateResponse(key, rsp); err != nil {
+		return fmt.Errorf("invalid response: %w", err)
+	}
+
+	// Update local state.
+	r.state[key] = value
+
+	return nil
+}
+
+func (r *runtime) doGetRequest(ctx context.Context, rng *rand.Rand, rtc runtimeClient.RuntimeClient, existing bool) error {
+	key := r.generateVal(rng, existing)
+
+	// Submit request.
+	rsp, err := r.submitRuntimeRquest(ctx, rtc, &runtimeTransaction.TxnCall{
+		Method: "get",
+		Args:   key,
+	})
+	if err != nil {
+		return fmt.Errorf("submit get request failed: %w", err)
+	}
+	r.logger.Debug("get request success",
+		"response", rsp,
+	)
+
+	if err := r.validateResponse(key, rsp); err != nil {
+		return fmt.Errorf("invalid response: %w", err)
+	}
+
+	return nil
+}
+
+func (r *runtime) doRemoveRequest(ctx context.Context, rng *rand.Rand, rtc runtimeClient.RuntimeClient, existing bool) error {
+	key := r.generateVal(rng, existing)
+
+	// Submit request.
+	rsp, err := r.submitRuntimeRquest(ctx, rtc, &runtimeTransaction.TxnCall{
+		Method: "remove",
+		Args:   key,
+	})
+	if err != nil {
+		return fmt.Errorf("submit remove request failed: %w", err)
+	}
+	r.logger.Debug("remove request success",
+		"response", rsp,
+	)
+
+	if err := r.validateResponse(key, rsp); err != nil {
+		return fmt.Errorf("invalid response: %w", err)
```

Comment:
some of these are missing context info:
- what kind of request it was
- info about the request (should we?)

Reply:
Fixed, also elsewhere. Added log messages that log the request, response, error.